docs/design/single-file-unsupported.md

```diff
@@ -0,0 +1,68 @@
+# Single-file unsupported attribute
+
+Starting with .NET Core 3.0 users have the ability to bundle all files used by their applications into a single executable through the publish functionality, `PublishSingleFile`. Although most applications can be agnostic of whether they will be deployed as single-file, there exist some APIs that are known to be incompatible with this deployment mode and thus can cause single-file apps to crash.
+
+While the existing single-file analyzer can help us emit warnings whenever incompatible runtime library APIs are called from within user's code in a project that has been marked with the intent of being published as single-file (through setting `PublishSingleFile` to true in its .csproj), developers can still take a dependency on third-party components implementing functionality that inadvertently make their application no longer compatible with single-file.
+
+## User Experience
+
+For library authors, the new attribute will allow them to easily annotate APIs known to be problematic when used in the context of a single-file application.
+
+```C#
+[SingleFileUnsupported("'Bar' method is not compatible with single-file", "https://help")]
+void Bar()
+{
+    Assembly executingAssembly = Assembly.GetExecutingAssembly();
+    var codeBase = executingAssembly.CodeBase;
+    ...
+}
+```
+
+Consumers of this library who want to publish their application as a single-file binary will commonly do it through setting the corresponding property in the application’s .csproj:
+
+```XML
+<PropertyGroup>
+    <PublishSingleFile>true</PublishSingleFile>
+</PropertyGroup>
+```
+
+If the user now makes use of the problematic API, a warning will be produced:
```

Comment:
How is that possible? I'm writing a library that references another library that has SingleFileUnsupported on its API. Where is the library author going to see the warning?

Reply:
I would maybe rewrite the text here to not talk about libraries directly, just code in general. As in "one method calls another method".
The libraries/apps distinction and experiences should be mentioned as well, but they're not directly related to the exact mechanics of how the attribute works.

Reply:
> How is that possible? I'm writing a library that references another library that has SingleFileUnsupported on its API. Where is the library author going to see the warning?

If the API that the library author is calling is annotated with the new attribute, the single-file analyzer will show a warning that should be visible in the IDE. Unfortunately, the author will not see any warnings for cases where the marked method is called indirectly (at least not until we have a global analyzer for this scenarios).

Foo:
```C#
[SingleFileUnsupported]
A() { }

B() { A(); }
```

My app:
```C#
using Foo;

// Analyzer will not be able to see the call to A and warn
C() { B(); }

// A warning should be displayed
D() { A(); }
```

Reply:
> the single-file analyzer will show a warning that should be visible in the IDE

How are you going to achieve that? Every developer would have to edit their project and add an unknown custom analyzer, then enable a single-file rule to warn, who is going to do that?

Reply:
I don't have an answer for libraries...
For apps, the idea is that specifying `PublishSingleFile=true` will enable the analyzer - no need to add it to the project, just add that property to the project. This is actually already the case in .NET 5 with the existing single-file Roslyn analyzer.

This would also work for libraries, but I agree that the property is somewhat confusing and not right for libraries, so we may need a different approach.

Reply:
You still need to include the linker analyzer dll somewhere in the project or is it already referenced by default?

Reply:
The Roslyn analyzer is always "included" - it just disables itself based on some properties in MSBuild - no user interaction necessary. I would image we would do something similar for the global analyzer (be it linker or something else).